README.md

# Gameserver
![Discord Shield](https://discordapp.com/api/guilds/1101926320465772574/widget.png?style=shield)

**Gameserver** is a centralized server specifically designed for **Wii**, providing efficient management of a diverse range of essential services, including JMCS, WDF, and Galaxy. This powerful server acts as the core infrastructure, responsible for coordinating and overseeing these critical components, ensuring optimal performance and seamless integration within the DanceParty ecosystem.

It's important to note that the Gameserver should not be confused with our repository, **KrystalUDP**, which focuses on handling UDP connections for online services of PlayStation 3, Xbox 360, and Wii U. The Gameserver and KrystalUDP serve distinct purposes and cater to different platforms within our comprehensive online infrastructure.

## Games & features availability
| Game |  World Dance Floor | Leaderboards | Online Challengers | Streamable Maps | Monthly Mashups  |
| ------------ | ------------ | ------------ | ------------ | ------------ | ------------ |
| 2014 | Under work | - | - | - |   Available |
| 2015 | Available | Available | Under work | Under work | - |
| 2016 | Available | Available | Under work | Under work | - |
| 2017 | Available | Available | Under work | Under work | - |
| 2018 | Available | Available | Under work | Under work | - |

## Sub-services

- **JMCS** (Jean Mich Central Server):
JMCS takes charge of numerous essential services, including leaderboards, user profiles, online maps, and the JD Wall. It serves as the core backend system responsible for managing and facilitating these functionalities, enhancing the overall user experience.

- **WDF** (World Dance Floor):
WDF is responsible for the management of various World Dance Floor data, encompassing playlists, status updates, and parties. It serves as the central hub for handling and organizing these crucial elements of the World Dance Floor experience.

- **NAS** (Nintendo Authentication Server):
NAS handles authentication tokens from the Wii console. We developed a custom NAS instead of using Wiimmfi due to the limitations of their token payload, which did not meet the requirements of our system.

- **Tracking**:
Tracking plays a crucial role in identifying issues, bugs, and gathering information sent by the game. However, we currently lack detailed information on how the API of Tracking is intended to function, which hinders its optimal performance. As a result, the tracking functionality may not be operating at its full capacity and may require further development and refinement.

- ~~**Galaxy**~~:
Galaxy was initially designed to handle the distribution and storage of the song database for both JMCS and WDF. However, we have made the decision to remove Galaxy from the repository as it is no longer in active use. Alternative solutions have been implemented to fulfill the role that Galaxy previously served, ensuring efficient song management and storage within JMCS and WDF.

### Requirements

- Node.JS *(version 18 or higher)*
- PM2 *(install via NPM)*
	### Databases
	- MongoDB
	- Memcached
  
### Installation and Usage

To install and use the Gameserver, follow these steps:

```shell
git clone https://github.com/DancePartyOnline/gameserver.git
cd gameserver
npm install
pm2 start ecosystem.json
```

The above code block will clone the server repository, install all required modules, and start all services in `ecosystem.json` with PM2.

### Using the Command Line

If you prefer not to run the server with PM2 or want to run a specific service, you can use the command line interface.

```shell
gameserver.js <command>

Commands:
	gameserver.js serve <service> Initialize a service

Options:
	--help Show help [boolean]
	--version Show version number [boolean]
	-e, --env Service environment [string]
	-p, --port Port to bind on [number]
```

### Examples

- `node src\gameserver.js serve jmcs --env local`: Start JMCS in the local environment.
- `node src\gameserver.js serve wdf --env dev --port 5000`: Start WDF in the development environment, binding to port 5000.

## Changelog

For a detailed list of changes, refer to our [changelog](https://github.com/dancepartyonline/gameserver/blob/main/CHANGELOG.md).

## Contribute

We welcome contributions to enhance our services, solutions, and workflow. Feel free to contribute!


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.8.7] - 2023-08-07

### Added
- Database migration, songDB is now handled by migrate-db. If you edit, create or delete songs, migrate-db will automatically detect changes and it will upsert the mongoDB database.
- NGINX configurations for services are now updated each server restart, we created this because we have multiple environments for 5+ services and it's painful to update all of them regularly. You can enable them with ENV value, e.g. `NGINX_LIB_ENABLED=true`
- With the new NGINX library, you can declare path to your sites-enabled/sites-available with ENV value, e.g. `NGINX_SITES_PATH=/etc/nginx/sites-enabled`
### Updated
- DB models
- Cache library

## [1.8.6] - 2023-08-05

### Added
- Structure for the new Shop service
- Template folder for services
- Library to connect to MGMT, to fetch shop songs
### Updated
- Configuration

## [1.7.6] - 2023-07-11

### Added
## [WDF API] 
- Added Dancercard router, with a total dancercard count route
- Total song count route to Songs router

## [1.7.5] - 2023-07-10

### Added
- Swagger documentation for WDF API, available at the `/api/swagger` route.
- With `gameserver/src/services/wdf/api/generate-docs-swagger.sh`, the script can be used to generate swagger documentation for the API.

### Fixed
- A misconfiguration that led API routes to be accessed without API token.

## [1.7.4] - 2023-07-10

### Added
- Documentation for WDF API, available at the `/api` route.
- With `gameserver/src/services/wdf/api/generate-docs.sh`, the script can be used to generate documentation for the API.

### Updated
- Gameserver global variables

## [1.7.3] - 2023-06-21

### Updated
- Request validation schemas
- Name regex

### Fixed
- Issue with request validation
- "connectToWDF" now turns everyones name uppercase, which means names of users who edit their save files to have lowercase will appear uppercase to any other player.
- Session library's randomSession function's first parameter, count now has Number(), which threw error for the first fix mentioned above.

## [1.7.2] - 2023-06-21

### Updated
- Updated games config and WDF room names

## [1.7.1] - 2023-06-20

### Added
- Updated `gameserver.js` port listener, now services can listen to HTTPS port with valid certificates.
- Added a new variable `FQDN` to service specific config to identify the service's domain. Future updates might update Security wall to verify the FQDN.
- Added new variables for service specific config: `SSL_PK`, `SSL_CERT`, `SSL_CA` which needs to be full path to the certificates of the service.
- Added a new variable `IS_ON_CLOUDFLARE` to service specific config (or Gameserver config to affect all services), to determine whether Gameserver is proxied through Cloudflare or not. It is used by security wall, used to get client's country from Cloudflare's `cf-country` header to check if the client's country is banned from accessing the services. When server is used without Cloudflare, security wall threw and error on every request indicating that there is no header for the country checked. This new config skips this check.

### Updated
- NAS's `ac` router error handler now gives more information on the error from Wiimmfi and gives 401 for Dolphin NANDs.

### Removed
- Removed Russia and Belarus from blocked countries.

config/index.js
// Fetched from Dolphin Emulator. Used by JMCS, WDF and DLC store 
// for client's country/region and localization.

module.exports.ENV = process.env.ENV || "local";

module.exports.LANGS = ["JA", "EN", "DE", "FR", "ES", "IT", "NL", "ZH", "ZH", "KO"];
module.exports.LANG_IDS = ["00", "01", "02", "03", "04", "05", "06", "07", "08", "09"];
module.exports.REGIONS = ["NTSC_J", "NTSC", "PAL", "NTSC_K"];
module.exports.REGION_IDS = ["00", "01", "02", "03"];

// Set database config by service and env
module.exports.DATABASE = require("./database")[global.service.id][global.env];

// Services
module.exports.SERVICES = {
    galaxy: {
        id: "galaxy",
        name: "Galaxy",
        path: "/services/galaxy/server.js",
    },
    jmcs: {
        id: "jmcs",
        name: "JMCS",
        path: "/services/jmcs/server.js"
    },
    wdf: {
        id: "wdf",
        name: "World Dance Floor (WDF)",
        path: "/services/wdf/server.js"
    }
};